File: api/src/sim/serialize.ts
/**
 * Canonical text for a battle result.
 *
 * The determinism requirement is byte-identical *state and events*, which needs
 * one agreed rendering: two runs are compared by comparing this text. It is also
 * what the headless demo prints, so a fresh process can be checked against an
 * in-process run without a second format existing.
 *
 * Line-delimited JSON, with the survivors sorted by id — array order is an
 * implementation detail of how units died, and comparing it would fail runs that
 * are in fact identical.
 */
import type { BattleResult } from './runBattle.js';

function eventLine(event: BattleResult['events'][number]): string {
  return JSON.stringify({
    event: event.type,
    tick: event.tick,
    position: { x: event.position.x, y: event.position.y },
    actors: {
      source: event.actors.source?.unitId ?? null,
      targets: event.actors.targets.map((target) => target.unitId),
    },
    swing: {
      zoneScore: event.swing.zoneScore,
      baseHp: event.swing.baseHp,
      unitsRemoved: event.swing.unitsRemoved.map((unit) => unit.unitId),
    },
  });
}

export function serializeBattle(result: BattleResult): string {
  const header = JSON.stringify({
    seed: result.seed,
    map: result.map.id,
    tickRate: result.config.tickRate,
    outcome: result.outcome,
    ticks: result.finalState.tick,
  });

  const survivors = [...result.finalState.units]
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0))
    .map((unit) => ({
      id: unit.id,
      side: unit.side,
      typeId: unit.typeId,
      hp: unit.hp,
      position: { x: unit.position.x, y: unit.position.y },
    }));

  const footer = JSON.stringify({
    finalUnits: survivors,
    bases: { north: result.finalState.bases.north.hp, south: result.finalState.bases.south.hp },
    zoneScore: result.finalState.zoneScore,
  });

  return [header, ...result.events.map(eventLine), footer].join('\n');
}

/**
 * A short fingerprint of a battle, for comparing runs at a glance.
 *
 * FNV-1a, written out rather than taken from `node:crypto`: the sim imports
 * nothing from Node, and `purity.test.ts` enforces that.
 */
export function digestBattle(result: BattleResult): string {
  const text = serializeBattle(result);
  let hash = 0x811c9dc5;

  for (let index = 0; index < text.length; index += 1) {
    hash ^= text.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193);
  }

  return (hash >>> 0).toString(16).padStart(8, '0');
}

File: api/app/sim/serialize.py
"""Canonical text for a battle result.

The determinism requirement is byte-identical *state and events*, which needs one
agreed rendering: two runs are compared by comparing this text. It is also what
the headless demo prints, so a fresh process can be checked against an in-process
run without a second format existing.

Line-delimited JSON, with the survivors sorted by id — list order is an
implementation detail of how units died, and comparing it would fail runs that
are in fact identical.
"""

from __future__ import annotations

import json

from app.sim.run_battle import BattleResult

MASK32 = 0xFFFFFFFF


def _event_line(event) -> str:  # type: ignore[no-untyped-def]
    return json.dumps(
        {
            "event": event.type,
            "tick": event.tick,
            "position": {"x": event.position.x, "y": event.position.y},
            "actors": {
                "source": event.actors.source.unit_id if event.actors.source else None,
                "targets": [target.unit_id for target in event.actors.targets],
            },
            "swing": {
                "zoneScore": dict(event.swing.zone_score),
                "baseHp": dict(event.swing.base_hp),
                "unitsRemoved": [unit.unit_id for unit in event.swing.units_removed],
            },
        },
        sort_keys=False,
        separators=(",", ":"),
    )


def serialize_battle(result: BattleResult) -> str:
    header = json.dumps(
        {
            "seed": result.seed,
            "map": result.map.id,
            "tickRate": result.config.tick_rate,
            "outcome": result.outcome,
            "ticks": result.final_state.tick,
        },
        separators=(",", ":"),
    )

    survivors = [
        {
            "id": unit.id,
            "side": unit.side,
            "typeId": unit.type_id,
            "hp": unit.hp,
            "position": {"x": unit.position.x, "y": unit.position.y},
        }
        for unit in sorted(result.final_state.units, key=lambda unit: unit.id)
    ]

    footer = json.dumps(
        {
            "finalUnits": survivors,
            "bases": {
                "north": result.final_state.bases["north"].hp,
                "south": result.final_state.bases["south"].hp,
            },
            "zoneScore": dict(result.final_state.zone_score),
        },
        separators=(",", ":"),
    )

    return "\n".join([header, *(_event_line(event) for event in result.events), footer])


def digest_battle(result: BattleResult) -> str:
    """A short fingerprint of a battle, for comparing runs at a glance.

    FNV-1a, written out rather than taken from `hashlib`: the sim imports nothing
    outside itself, and `tests/test_purity.py` enforces that.
    """
    text = serialize_battle(result)
    value = 0x811C9DC5

    for char in text:
        value ^= ord(char) & 0xFF
        value = (value * 0x01000193) & MASK32

    return format(value, "08x")
